### PR TITLE
feat(runner): Integrate runner for Jenkins jobs

### PR DESCRIPTION
**Pull request by buccarel**
Fixes the following issue:
https://github.com/chainloop-dev/chainloop/issues/88

Definition of done:
[x] Implementation of the runner
[x] Unit tests
[x] Local tests

Jenkins output of `att init`

image

Generated chainloop_attestation.tmp.json

image

### CODE REVIEW

File: internal/attestation/crafter/runners/jenkinsjob.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package runners
+
+import "os"
+
+type JenkinsJob struct{}
```
Comment (migmartri) [inline internal/attestation/crafter/runners/jenkinsjob.go:20]: This is not related to your change, since the other runners have the same issue, but it's something that we could tackle in another patch.

It seems that the "factory functions" are not being used outside of tests, when they could also be used in `runner.go` in the switch clause.

File: app/controlplane/api/workflowcontract/v1/crafting_schema.pb.go [hunk 2]
```diff
@@ -43,6 +43,7 @@ const (
 	CraftingSchema_Runner_GITHUB_ACTION           CraftingSchema_Runner_RunnerType = 1
 	CraftingSchema_Runner_GITLAB_PIPELINE         CraftingSchema_Runner_RunnerType = 2
 	CraftingSchema_Runner_AZURE_PIPELINE          CraftingSchema_Runner_RunnerType = 3
+	CraftingSchema_Runner_JENKINS_JOB	      CraftingSchema_Runner_RunnerType = 4
```
Comment (migmartri) [inline app/controlplane/api/workflowcontract/v1/crafting_schema.pb.go:46]: It seems that you've manually updated this value when in fact this is the result of modifying crafting_schema.proto

https://github.com/chainloop-dev/chainloop/blob/01432f04375e2d6a25084fe88c29eff1c32f5da4/app/controlplane/api/workflowcontract/v1/crafting_schema.proto#L42-L46

and then running `make api`. 

The good news is that I believe you've found a bug, for some reason our CI has not detected this, and it should have. We have similar checks for other generated code but not this one interestingly.

Just for runs, if you run `make api` in the code of your patch, you'll see how this line gets removed

```diff
diff --git a/app/controlplane/api/workflowcontract/v1/crafting_schema.pb.go b/app/controlplane/api/workflowcontract/v1/crafting_schema.pb.go
index 3de6c8f..0b782cf 100644
--- a/app/controlplane/api/workflowcontract/v1/crafting_schema.pb.go
+++ b/app/controlplane/api/workflowcontract/v1/crafting_schema.pb.go
@@ -43,7 +43,6 @@ const (
        CraftingSchema_Runner_GITHUB_ACTION           CraftingSchema_Runner_RunnerType = 1
        CraftingSchema_Runner_GITLAB_PIPELINE         CraftingSchema_Runner_RunnerType = 2
        CraftingSchema_Runner_AZURE_PIPELINE          CraftingSchema_Runner_RunnerType = 3
-       CraftingSchema_Runner_JENKINS_JOB             CraftingSchema_Runner_RunnerType = 4
 )

```

So as I said, instead, modify the proto definition and run the generation, that should persist it.

Thanks!

Comment (buccarel) [reply app/controlplane/api/workflowcontract/v1/crafting_schema.pb.go:46]: Ohh interesting!
I bet it's something I know, like some knowledge I am missing, because I didn't realize at all that file was generated.
My bad my bad.
Will fix it!
Thanks

**Comment by migmartri** [comment]
To run this locally, since this is just a client-side update you have two options:

a) run full development environment including CAS, controlplane locally or
b) just run your CLI pointing to the already existing SASS chainloop instance

For the latter, and to perform an attestation, even locally, you need to configure a workflow + robot-account and this is how you do it.


- login to the controlplane `go run main.go auth login`
- Create an empty contract (github.yaml) that sets the new runner requirement

```yaml
schemaVersion: v1

runner:
  type: "JENKINS_JOB"
```

- create a workflow pointing to that empty contract, that should return a robot-account. This is an API token that's used during attestation.

```console
go run main.go wf create --name test --project jktest --contract jenkins.yaml
```

Export the robot-account as an env variable

```sh
export CHAINLOOP_ROBOT_ACCOUNT=REDACTED.REDACTED.tBAf4SzsCWjFzuaS9R6a0PBNSo2_wdPjZKH1x_ogObc
```

and then run an `attestation init`, you should get the expected error

```console
$ go run main.go att init                                                           
ERR The contract expects the attestation to be crafted in a runner of type "jenkins-job" but couldn't be detected. Use --dry-run flag if development
```

after this you know how to proceed, impersonating the env vars or even running the modified version of this CLI (compiled with `make -C app/cli build` in a Jenkins runner.

I hope it helps

File: app/controlplane/api/workflowcontract/v1/crafting_schema.pb.go (hunk 2 above)
Comment (migmartri) [reply app/controlplane/api/workflowcontract/v1/crafting_schema.pb.go:46]: as I said, the CI should have warned us, I'll fix that :)

Comment (migmartri) [reply app/controlplane/api/workflowcontract/v1/crafting_schema.pb.go:46]: @buccarel, this change https://github.com/chainloop-dev/chainloop/pull/412 is what was missing in the CI to detect your use-case.

If we happen to merge it and your code rebased, your PR should then fail telling you that you are not committing all the generated code or that it differs from the baseline.

**Comment by migmartri** [comment]
#412 has been merged, please rebase main now, thanks!

**Review by migmartri — APPROVED** [review]
LGTM! 

Next we'll need to update the docs https://docs.chainloop.dev/reference/operator/contract#runner-context but we can wait until they are moved to the main repo. @danlishka is working on that

Thanks!
